mitmproxy/proxy/protocol2/websocket.py

```diff
@@ -16,14 +19,104 @@ def __init__(self, context: ClientServerContext, handshake_flow: http.HTTPFlow):
         super().__init__(context)
         self.flow = websocket.WebSocketFlow(context.client, context.server, handshake_flow)
         assert context.server.connected
+        self.client_frame_buffer = []
+        self.server_frame_buffer = []
+        self.extension = None
+        self.make_extension()
+        self.client_conn = wsconn.WSConnection(wsconn.SERVER, wsconn.ConnectionState.OPEN,
+                                               extensions=[self.extension] if self.extension else None)
+
+        self.server_conn = wsconn.WSConnection(wsconn.CLIENT, wsconn.ConnectionState.OPEN,
+                                               extensions=[self.extension] if self.extension else None)
+
+    def make_extension(self):
+        if self.flow.server_extensions:
+            name, *params = self.flow.server_extensions.split(';')
+            if name == 'permessage-deflate':
+                args = {
+                    "client_no_context_takeover": False,
+                    "server_no_context_takeover": False,
+                    "client_max_window_bits": None,
+                    "server_max_window_bits": None
+                }
+
+                for param in params:
+                    if param.startswith('client_no_context_takeover'):
+                        args["client_no_context_takeover"] = True
+                    elif param.startswith('server_no_context_takeover'):
+                        args["server_no_context_takeover"] = True
+                    elif param.startswith('client_max_window_bits'):
+                        args["client_max_window_bits"] = int(param.split('=', 1)[1].strip())
+                    elif param.startswith('server_max_window_bits'):
+                        args["server_max_window_bits"] = int(param.split('=', 1)[1].strip())
+
+                self.extension = PerMessageDeflate(**args)
```

Comment:
I think we should move parts of this into wsproto itself, and provide a simple API (to be used by mitmproxy here) to access these parameters.
Ideally IMO we shouldn't deal with extensions at all, but simply pass the completed handshake back to wsproto to "update" it's internal state.